compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -198,15 +199,68 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 let (source, ty) = if let ty::Adt(adt_def, ..) = source.ty.kind() && adt_def.is_enum() {
                     let discr_ty = adt_def.repr().discr_type().to_ty(this.tcx);
                     let temp = unpack!(block = this.as_temp(block, scope, source, Mutability::Not));
+                    let layout = this.tcx.layout_of(this.param_env.and(source.ty));
                     let discr = this.temp(discr_ty, source.span);
                     this.cfg.push_assign(
                         block,
                         source_info,
                         discr,
                         Rvalue::Discriminant(temp.into()),
                     );
+                    let (op,ty) = (Operand::Move(discr), discr_ty);
+
+                    if let Abi::Scalar(scalar) = layout.unwrap().abi{
+                        if let Primitive::Int(..) = scalar.primitive(){
+                            let range = scalar.valid_range(&this.tcx);
+                            if range.end >= range.start {
+                                let range_end =
```

Comment:
You showed me tests on zulip that I think would require sign extension here, too. Please add these tests.

To simplify this logic here, maybe put it into a closure and call it twice, once for start and once for end. That way the logic only exists once and you avoid such bugs.